### PR TITLE
[vsphere] support hardware version 20 on vsphere8

### PR DESCRIPTION
**Pull request by jinyunma**
support hardware version 20 on vsphere8

### CODE REVIEW

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-vsphere-8-ovn-upi

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.11-e2e-vsphere-upi

**Comment by jinyunma** [comment]
cc @jianlinliu to review

From upi-conf-vsphere-ova [build-log.txt](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/36910/rehearse-36910-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere/1631519463314558976/artifacts/installer-rehearse-vsphere/upi-conf-vsphere-ova/build-log.txt), checked that ova image is cloned and upgraded to hw20 successfully

```
# govc vm.info -json /IBMCdatacenter/vm/rhcos-413.86.202302150245-0-vmware.x86_64.ova-hw20  | jq -r .VirtualMachines[].Config.Version
vmx-20
```

File: ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh [hunk 1]
```diff
@@ -70,8 +71,12 @@ for i in "${!DATACENTERS[@]}"; do
         wait "$!"
     fi
 
-    hw_versions=(15 17 18 19)
+    hw_versions=(15 17 18 19 20)
```
Comment (jianlinliu) [inline ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh:74]: I read the logic, seem like it is uploading ova with hw version in order, can you help point me how did we randomly pick up hw version in the testing?

File: ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh [hunk 2]
```diff
@@ -54,6 +54,7 @@ fi
 
 echo "$(date -u --rfc-3339=seconds) - Checking if RHCOS OVA needs to be downloaded from ${ova_url}..."
 
+vsphere_version=$(govc about -json | jq -r .About.Version | awk -F'.' '{print$1}')
```
Comment (jianlinliu) [inline ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh:57]: how about adding a whitespace between `print` and `$1`

File: ci-operator/step-registry/upi/conf/vsphere/upi-conf-vsphere-commands.sh [hunk 3]
```diff
@@ -91,6 +81,20 @@ source "${SHARED_DIR}/vsphere_context.sh"
 # shellcheck source=/dev/null
 source "${SHARED_DIR}/govc.sh"
 
+# select a hardware version for testing
+vsphere_version=$(govc about -json | jq -r .About.Version | awk -F'.' '{print$1}')
```
Comment (jianlinliu) [inline ci-operator/step-registry/upi/conf/vsphere/upi-conf-vsphere-commands.sh:85]: how about adding a whitespace between `print` and `$1`

File: ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh (hunk 2 above)
Comment (jinyunma) [reply ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh]: Updated

File: ci-operator/step-registry/upi/conf/vsphere/upi-conf-vsphere-commands.sh (hunk 3 above)
Comment (jinyunma) [reply ci-operator/step-registry/upi/conf/vsphere/upi-conf-vsphere-commands.sh]: updated

**Comment by jinyunma** [comment]
@jcpowermac could you help to review it when you get a chance? thanks.

**Comment by jcpowermac** [comment]
@jinyunma  we are running 15-19 HW versions in our other vSphere enviornments. Maybe for 8 we should just test 20?
@rvanderp3 wdyt?

the hardware version "configuration" (which we don't really know exactly what that is) is independant of ESXi version. Meaning if I have a v17 HW guest the "hardware" is the same on ESXi 7 or 8.

e.g. the vcsa (vcenter) appliance still uses a hardware version of 10.

File: ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh [hunk 4]
```diff
@@ -70,13 +71,17 @@ for i in "${!DATACENTERS[@]}"; do
         wait "$!"
     fi
 
-    hw_versions=(15 17 18 19)
+    hw_versions=(15 17 18 19 20)
     for hw_version in "${hw_versions[@]}"; do
```
Comment (jcpowermac) [inline ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh:75]: maybe I wasn't paying attention to a previous pr
but does this import an ova per datacenter per version? just make sure I am parsing that right

**Comment by jianlinliu** [comment]
> @jinyunma we are running 15-19 HW versions in our other vSphere enviornments. Maybe for 8 we should just test 20? @rvanderp3 wdyt?
> 
> the hardware version "configuration" (which we don't really know exactly what that is) is independant of ESXi version. Meaning if I have a v17 HW guest the "hardware" is the same on ESXi 7 or 8.
> 
> e.g. the vcsa (vcenter) appliance still uses a hardware version of 10.

A benefit to cover 15-20 HW versions in vsphere-8, is that we cover compatibility testing even there is some difference between different of vsphere deployment. Maybe we can be more confident when facing customers. :)

Another benefit, in the future, if we retired vsphere-7, move most of test jobs to vpshere-8, we do not need to update codes any more.

**Comment by jcpowermac** [comment]
> A benefit to cover 15-20 HW versions in vsphere-8, is that we cover compatibility testing even there is some difference between different of vsphere deployment that we might do not know. So we can be more confident when facing customers. :)

I don't agree with that statement. Again we are already testing 15-19 in vSphere 7. HW version is irrespective of ESXi version in its configuration - VMware can't change it between ESXi releases.
You would just be testing HW20 less often and have less confidence in vSphere 8+HW20.

> Another benefit, in the future, if we retired vsphere-7, move most of test jobs to vpshere-8, we do not need to update codes any more.

its a couple lines of bash, I think we could manage

**Comment by jianlinliu** [comment]
> Again we are already testing 15-19 in vSphere 7. HW version is irrespective of ESXi version in its configuration - VMware can't change it between ESXi releases. You would just be testing HW20 less often and have less confidence in vSphere 8+HW20.

If that, I am okay with it, especially we still have vpshere 7 to run regular ci jobs. 

Let us see if someone have some other opinions.

**Comment by jinyunma** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

**Comment by jinyunma** [comment]
@jcpowermac @jianlinliu changed to only test HW20 on vsphere 8, please help to review again, thanks.

File: ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh (hunk 4 above)
Comment (jinyunma) [reply ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh:78]: yeah, to test vsphere UPI zonal deployment, see https://github.com/openshift/release/pull/35175/

File: ci-operator/step-registry/upi/conf/vsphere/upi-conf-vsphere-commands.sh [hunk 5]
```diff
@@ -91,6 +81,20 @@ source "${SHARED_DIR}/vsphere_context.sh"
 # shellcheck source=/dev/null
 source "${SHARED_DIR}/govc.sh"
 
+# select a hardware version for testing
+vsphere_version=$(govc about -json | jq -r .About.Version | awk -F'.' '{print $1}')
+if [[ ${vsphere_version} -lt 8 ]]; then
+    hw_versions=(15 17 18 19)
+else
+    hw_versions=(20)
+fi
```
Comment (jianlinliu) [inline ci-operator/step-registry/upi/conf/vsphere/upi-conf-vsphere-commands.sh:90]: How about keep the same logic like ci-operator/step-registry/upi/conf/vsphere/ova/upi-conf-vsphere-ova-commands.sh.
```
hw_versions=(15 17 18 19)
if [[ ${vsphere_version} -eq 8 ]]; then
        hw_versions=(20)
fi
```

Comment (jinyunma) [reply ci-operator/step-registry/upi/conf/vsphere/upi-conf-vsphere-commands.sh:89]: updated.

**Comment by jcpowermac** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-vsphere-8-ovn-upi

**Pull request merged**